### PR TITLE
Move snapshot tests to bazel

### PR DESCRIPTION
**Pull request by elliottt**
### New features:
* Test and update targets for snapshot tests
```
$ bazel test //gems/sorbet/test/snapshot
$ bazel test //gems/sorbet/test/snapshot:update
```
* Test discovery for tests in the `partial` and `total` sub-directories of `gems/sorbet/test/snapshot`
* Updates for snapshots are integrated with `tools/scripts/update_exp_files.sh`

### Caveats:
* `partial/stupidedi` takes 267s to finish
* `partial/webmock` and `partial/partially-sigged` require openssl support
* `partial/codecov` and `partial/rails-double-require` require gems that
  build native code
* New tests must have their Gemfile.lock files added to WORKSPACE
* `sorbet-typed` gets cloned a lot

### Motivation

This moves the snapshot testing into bazel, allowing them to be sandboxed.

### Test plan
The `//gems/sorbet/test/snapshot` test suite now runs under bazel.

### CODE REVIEW

File: tools/scripts/update_exp_files.sh [hunk 1]
```diff
@@ -4,12 +4,12 @@ set -e
 COMMAND_FILE=$(mktemp)
 trap 'rm -f "$COMMAND_FILE"' EXIT
 
-# TODO(JEZ) Make this faster and better so we can uncomment this
```
Comment (jez) [inline tools/scripts/update_exp_files.sh]: How much slower does this make update_exp_files.sh? I think that fundamentally `srb init` is slow regardless of bazel, and we probably still want to keep "updating testdata tests" and "updating srb init tests" separate.

Comment (elliottt) [reply tools/scripts/update_exp_files.sh]: It's slow, but everything is running in parallel, so it didn't feel slower than normal. I'll try a run on `master` and this branch, and see what the difference is.

File: gems/sorbet/test/snapshot/BUILD [hunk 2]
```diff
@@ -1,4 +1,45 @@
...
+
+snapshot_tests(
+    name = "snapshot",
+    tests = glob(
+        [
+            "partial/*",
+            "total/*",
+        ],
+        exclude = [
+            # openssl is missing
```
Comment (jez) [inline gems/sorbet/test/snapshot/BUILD]:
```suggestion
            # TODO(jez) openssl is missing
```

File: gems/sorbet/test/snapshot/BUILD [hunk 3]
```diff
@@ -1,4 +1,45 @@
...
+            "partial/webmock",
+            "partial/partially-sigged",
+
+            # native extensions don't build
```
Comment (jez) [inline gems/sorbet/test/snapshot/BUILD]:
```suggestion
            # TODO(jez) native extensions don't build
```

File: gems/sorbet/test/snapshot/check_one_bazel.sh [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+# Runs a single srb init test from gems/sorbet/test/snapshot/{partial,total}/*
+
+set -euo pipefail
+shopt -s dotglob
+
+source "gems/sorbet/test/snapshot/logging.sh"
+
+# ----- Option parsing -----
+
+test_name=$1
```
Comment (jez) [inline gems/sorbet/test/snapshot/check_one_bazel.sh]: For my own sanity, can we do some arg parsing at the top of the script?

- immediately check that we have the correct required number of args
- immediately assign positional args into variables with names
- ideally show a decent help message if something's wrong

I imagine that most of the time this won't matter because bazel will pass the right args, but it's still nice for the reader to get oriented.

Comment (jez) [reply gems/sorbet/test/snapshot/check_one_bazel.sh]: i.e., `$2` is hiding about 15 lines down

File: gems/sorbet/test/snapshot/check_one_bazel.sh [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+else
+  is_partial=
+fi
+
+info "├─ test_name: ${test_name}"
+info "├─ is_partial: ${is_partial:-0}"
+
+# ----- Environment setup -----
+
+repo_root=$PWD
```
Comment (jez) [inline gems/sorbet/test/snapshot/check_one_bazel.sh]: Idk if this affects your script, but `$PWD` is allowed to have symlinks in it, i.e., it's not the case that

```bash
[ "$PWD" = "$(realpath "$PWD")" ]
```

in all cases (again not sure if this is important for you).

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 6]
```diff
@@ -0,0 +1,176 @@
+#!/bin/bash
+
+# Runs a single srb init test from gems/sorbet/test/snapshot/{partial,total}/*
+
+shopt -s dotglob
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: Do you think it's possible to put `set -euo pipefail` down below the runfiles.bash dance?

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+if [[ "${test_name}" =~ partial/* ]]; then
+  is_partial=1
+else
+  is_partial=
+fi
+
+info "├─ test_name:  ${test_name}"
+info "├─ output_archive:  ${output_archive}"
+info "├─ is_partial: ${is_partial:-0}"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: Slight preference, can we make these vertically aligned?

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+# Add ruby to the path
+# shellcheck disable=SC1090
+PATH="${repo_root}/$(dirname "$(rlocation "${ruby_package}/ruby")"):$PATH"
+export PATH
+
+# Add bundler to the path
+# shellcheck disable=SC1090
+BUNDLER_LOC="${repo_root}/$(dirname "$(rlocation "gems/bundler/bundle")")"
+GEMS_LOC="$BUNDLER_LOC/../gems"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: in my experience having `..` in paths can get annoying to work with. I might have replaced this with another `dirname` call on `BUNDLER_LOC` instead of `../`. (non blocking)

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+export PATH
+
+test_dir="${repo_root}/gems/sorbet/test/snapshot/${test_name}"
+
+actual="${test_dir}/actual"
+
+srb="${repo_root}/gems/sorbet/bin/srb"
+
+# Use the sorbet executable built by bazel
+SRB_SORBET_EXE="$PWD/$(rlocation com_stripe_ruby_typer/main/sorbet)"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: nice

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+# shellcheck disable=SC1090
+PATH="${repo_root}/$(dirname "$(rlocation "${ruby_package}/ruby")"):$PATH"
+export PATH
+
+# Add bundler to the path
+# shellcheck disable=SC1090
+BUNDLER_LOC="${repo_root}/$(dirname "$(rlocation "gems/bundler/bundle")")"
+GEMS_LOC="$BUNDLER_LOC/../gems"
+PATH="$BUNDLER_LOC:$PATH"
+export PATH
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: ooc, why are you always exporting on a newline instead of `export PATH="..."`?

Comment (DarkDimius) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: shellcheck prohibits this

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+export XDG_CACHE_HOME
+
+info "├─ ruby: $(command -v ruby)"
+info "├─ ruby --version: $(ruby --version)"
+
+info "├─ bundle: $(command -v bundle)"
+info "├─ bundle --version: $(bundle --version)"
+
+info "├─ sorbet: $SRB_SORBET_EXE"
+info "├─ sorbet --version: $("$SRB_SORBET_EXE" --version)"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: What do you think about intermixing these with where they're initialized, so that if something ever breaks in the script between the last `info` and these, we can quickly pinpoint where it went wrong?

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+
+# ----- Build the test sandbox -----
+
+cp -r "${test_dir}/src" "$actual"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: Maybe it's just because I'm familiar with the old script, but I remember there used to be some annoying bits here to get local gems working. How does that work now? Who's in charge of populating `src/../gems`?

Comment (jez) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: oh, maybe they all just come in via `vendor` below?

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+
+# ----- Run the test -----
+
+(
+  cd "$actual"
+
+  # Setup the vendor/cache directory to include all gems required for any test
+  info "├─ Setting up vendor/cache"
+  mkdir vendor
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: idk if this can make a difference, but I always use `mkdir -p` which will not fail if `vendor/` already exists as a directory

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+  ln -s "$GEMS_LOC" "vendor/cache"
+
+  info "├─ Testing 'bundle exec which ruby'"
+  if bundle exec which ruby ; then
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]:
```suggestion
  if bundle exec which ruby; then
```

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+
+(
+  cd "$actual"
+
+  # Setup the vendor/cache directory to include all gems required for any test
+  info "├─ Setting up vendor/cache"
+  mkdir vendor
+  ln -s "$GEMS_LOC" "vendor/cache"
+
+  info "├─ Testing 'bundle exec which ruby'"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: Can we remove this message and just improve the "success" or "failure" message?

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+  if bundle exec which ruby ; then
+    info "├─ success"
+  else 
+    error "└─ failure"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: If there's a failure here, do we always expect that it's something wrong with the interaction between `vendor/cache` + the Gemfile ? if so / if not, could we make the `failure` message say something about maybe what to try next?

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 17]
```diff
@@ -0,0 +1,176 @@
...
+    exit 1
+  fi
+
+  # Configuring output to vendor/bundle
+  # Passing --local to never consult rubygems.org
+  # Passing --no-prune to not delete unused gems in vendor/cache
+  info "├─ Installing dependencies to vendor/bundle"
+  bundle config --local path vendor/bundle
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: How is path meaningful? If it's an actual file path, I'd prefer `./path`, if it's some sort of enum that `--local` accepts, I'd prefer `"path"`

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+  fi
+
+  # FIXME: Removing hidden-definitions in actual to hide them from diff output.
+  rm -rf "sorbet/rbi/hidden-definitions"
+
+  # Fix up the logs to not have sandbox directories present.
+
+  info "├─ Fixing up err.log"
+  err_filtered="$(mktemp)"
+  sed -e "s,${TMPDIR}[^ ]*/\([^/]*\),<tmp>/\1,g" \
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: You can do this with `sed`:

```bash
sed -i.bak ... && rm err.log.bak
```

which avoids having to create the temp yourself. It also doesn't affect this specific incantation, but it will preserve file permissions (owner, group, executable bit, etc.).

File: gems/sorbet/test/snapshot/run_one_bazel.sh [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+
+)
+
+(
+  cd "$test_dir"
+
+  info "├─ archiving results"
+
+  # archive the test
+  tar -cz actual/{sorbet,err.log,out.log} > "$output_archive"
```
Comment (jez) [inline gems/sorbet/test/snapshot/run_one_bazel.sh]: huh i've never seen `tar` used like like. TIL

File: gems/sorbet/test/snapshot/update_one_bazel.sh [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+      diff_total
+    else
+      info "├─ Performing partial diff"
+
+      # we know that there's an expected dir, as that is a pre-requesite for this
+      # test to be run by bazel
+      diff_partial
+    fi
+  else
+    attn "├─ missing expected/sorbet directory, skipping update"
```
Comment (jez) [inline gems/sorbet/test/snapshot/update_one_bazel.sh]: Can you say something here about what the workflow is for recording a new test? i.e., imagine that I just added a single `Gemfile` to a new folder `partial/foo`, and I want to populate the expected folder automatically. Right now, you can pass `--update --record` which will have the behavior that a partial test is treated like a total test for the sake of updating it only. After that, it's easy to go remove the files that aren't relevant to this test, and keep the ones that are.

iiuc, after this change, to replicate this workflow, someone actually has to create a new total test, run update, and then make it a partial test + remove things.

File: gems/sorbet/test/snapshot/snapshot.bzl [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+    res["test_name"] = test_name
+
+    # Generate an update rule if the test has an expected directory
+    expected = native.glob(
+        ["{}/expected".format(test_path)],
+        exclude_directories = 0,
+    )
+
+    if len(expected) > 0:
```
Comment (jez) [inline gems/sorbet/test/snapshot/snapshot.bzl]: wow is this the only way you can check whether a file exists in bazel 🤔 they're really trying to make escaping the sandbox hard

File: gems/sorbet/test/snapshot/snapshot.bzl [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+            deps = [":logging"],
+            srcs = ["update_one_bazel.sh"],
+            args = [
+                "$(location {})".format(actual),
+                "$(location {}/expected)".format(test_path),
+                test_path,
+            ],
+
+            # NOTE: these tags cause this test to be skipped by `bazel test //...`,
+            # and not run in the sandbox.
```
Comment (jez) [inline gems/sorbet/test/snapshot/snapshot.bzl]: Can you put the comment for which is which on top of the individual tag? e.g., i don't think that `manual` has anything to do with the sandbox, but i'm not sure which of `external` / `local` does what.

File: gems/sorbet/test/snapshot/check_one_bazel.sh (hunk 4 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/check_one_bazel.sh]: Sure, I'll rework it to do the option parsing at the top instead.

File: gems/sorbet/test/snapshot/check_one_bazel.sh (hunk 5 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/check_one_bazel.sh]: `realpath` doesn't exist in the sandbox, so this was the best workaround that I had come up with.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 7 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: Yep, I'll align all the close groups of log messages.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 8 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: I've switched it to another use of `dirname`, thanks!

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 10 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: Yes, this is to avoid tripping that error when the shellcheck runs.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 11 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: I like that. I'll move them closer to where the PATH is being setup for each case.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 12 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: That works as it did before -- the Gemfile still refers to the same location, as `actual` is now just a mirror of `src`.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 13 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: The `vendor` dir should always be missing at this point, as the script will be run in a clean sandbox. I sort of like leaving it off, as it would be a good point of failure if we accidentally commit a vendor tree in `src`.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 15 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: Sounds good, I'll refactor the other cases of this as well.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 16 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: In this case, the problem would be related to `$PATH` not being setup correctly. I should also check the output to ensure that it's finding the ruby from the `@ruby_2_4_3` package, instead of the system ruby accidentally.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 17 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: This installs to `vendor/bundle` instead of `.bundle`. It's not super meaningful, so I'll try testing without it.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 18 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: Much better, thanks!

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 19 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: That's the way I learned it ages ago, and it's been a hard habit to break :)

File: gems/sorbet/test/snapshot/snapshot.bzl (hunk 21 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/snapshot.bzl]: There may be a better way than this. Wrapping this up as a custom rule would give access to a lot of other functionality, but given that i needed the file list anyway this seemed pretty reasonable :)

File: gems/sorbet/test/snapshot/snapshot.bzl (hunk 22 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/snapshot.bzl]: I've cleaned this up a bit, let me know if you'd like further documentation.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 17 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: It wasn't necessary. Since this is running in the sandbox, it won't matter where the installed gems go.

File: gems/sorbet/test/snapshot/update_one_bazel.sh (hunk 20 above)
Comment (elliottt) [reply gems/sorbet/test/snapshot/update_one_bazel.sh]: There isn't a great workflow right now. I think the best path forward is to add a script that builds the `actual.tar.gz` artifact for a test, and then manually moves things into the `expected` directory. Attempting to do this from within bazel will be difficult.

File: .buildkite/pipeline.yaml [hunk 23]
```diff
@@ -63,14 +63,6 @@ steps:
     <<: *elastic
   - wait: ~
 
-# Test Sorbet Gem Phase - Its own phase because it relies on the gem build built
```
Comment (DarkDimius) [inline .buildkite/pipeline.yaml]: why was this deleted?

Comment (elliottt) [reply .buildkite/pipeline.yaml]: It's moved to `test-static-sanitized.sh`, in the same test invocation as `@ruby_2_4_3` and `@gems`.

File: gems/sorbet/test/snapshot/run_one_bazel.sh (hunk 10 above)
Comment (jez) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: I believe shellcheck only prohibits it when you're running a command on the inside. So for example

```bash
export BUNDLER_LOC="${repo_root}/$(dirname "$(rlocation gems/bundler/bundle)")"
```

is not ok because there's a nested command, which has its own return code separate from the `export` statement's return code. But for 

```bash
export PATH="$BUNDLER_LOC:$PATH"
```

which is just joining two strings, there's no nested command so it's fine.

I'm fine keeping it as it is—I just wanted to point out that I'm pretty sure this is not a shellcheck restriction.

Comment (elliottt) [reply gems/sorbet/test/snapshot/run_one_bazel.sh]: Ah, thanks for clearing that up!

**Comment by elliottt** [comment]
I've added these issues to track follow on work:
* Update/create test workflow needs work: #1221 
* Debugging snapshot tests needs work: #1222
